lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+#include "dynamatic/Transforms/HandshakeOptimization.h"
+#include "dynamatic/Dialect/Handshake/HandshakeCanonicalize.h"
+#include "dynamatic/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Support/LLVM.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+#include <vector>
+using namespace mlir;
+using namespace dynamatic;
+
+namespace {
+
+struct RemoveBranchMUXPairs
+    : public OpRewritePattern<handshake::ConditionalBranchOp> {
+  using OpRewritePattern<handshake::ConditionalBranchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::ConditionalBranchOp condBranchOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner1 = nullptr;
+    mlir::Operation *useOwner2 = nullptr;
+
+    auto it = condBranchOp.getTrueResult().getUsers().begin();
+    if (it != condBranchOp.getTrueResult().getUsers().end()) {
+      useOwner1 = *it;
+    }
+
+    it = condBranchOp.getFalseResult().getUsers().begin();
+    if (it != condBranchOp.getFalseResult().getUsers().end()) {
+      useOwner2 = *it;
+    }
+
+    if (isa_and_nonnull<handshake::MuxOp>(useOwner1) and
+        useOwner1 == useOwner2) {
+      handshake::MuxOp muxOp = mlir::cast<handshake::MuxOp>(useOwner1);
+      mlir::Value conditionBr = condBranchOp.getConditionOperand();
+      mlir::Value selectMux = muxOp.getSelectOperand();
+
+      if (conditionBr.getDefiningOp() == selectMux.getDefiningOp() and isa_and_nonnull<handshake::ForkOp>(conditionBr.getDefiningOp())) {
```

Comment:
By the IR's materialization property if both operands have the same defining operation then it is *necessarily* a fork so the second part of that condition is implied.

I am going to stop commenting on this particular materialization or not issue and we should chat offline about what are the actual requirements of this pass, because us being unclear on that is actually significantly hurting the implementation.